Typescript and Rehype refactor

ChristianMurphy opened the pull request:
Unified, through remark and rehype is providing the translation of markdown source into react elements.
TypeScript is used to allow for catching type related bugs and automatic generation of typings files.
tsdx is mananging the build process.
Storybook is leveraged to create runnable examples which can be hosted on GitHub pages.
GitHub actions are used for continuous integration and automatic deployment of the demo site.

There is a live demo of this code running at https://christianmurphy.github.io/react-remark
and a scoped version available on the registry with `npm install --save @christianmurphy/react-remark`


resolves #427

ChristianMurphy commented on src/index.ts:
```diff
@@ -0,0 +1,69 @@
+} from 'react';
+import unified, { PluggableList } from 'unified';
+import remarkParse, { RemarkParseOptions } from 'remark-parse';
+// @ts-ignore
+import remarkToRehype from 'remark-rehype';
+// @ts-ignore
+import rehypeReact from 'rehype-react';
```
These packages don't have typings yet.
Work is being done upsteam to add typings.

ChristianMurphy commented on src/index.ts:
```diff
@@ -0,0 +1,69 @@
+}
+
+export const useRemark = ({
+  remarkParseOptions,
+  remarkToRehypeOptions,
+  rehypeReactOptions,
+  remarkPlugins = [],
+  rehypePlugins = [],
+  onError = () => {},
```
This API is different.
It may be possible to create an adapter which is more like the current API on top of this component to smooth the transition.

ChristianMurphy commented on stories/remark-hook.stories.tsx:
```diff
@@ -0,0 +1,44 @@
+  });
+  const markdownSource = text(
+    'markdown',
+    `Lift($L$) can be determined by Lift Coefficient ($C_L$) like the following equation.
+
+  $$
+  L = \\frac{1}{2} \\rho v^2 S C_L
+  $$`
+  );
```
This gives an example of embedding math in markdown, a use-case which leveraging rehype enables.

ChristianMurphy commented on test/remark-component.test.tsx:
```diff
@@ -0,0 +1,14 @@
+import '@testing-library/jest-dom/extend-expect';
+import { Remark } from '../src';
+
+describe('Remark', () => {
```
These tests focus on testing the React abstraction, remark-rehype and rehype-react have their own tests verifying content is converted as expected.

ChristianMurphy commented on .github/workflows/main.yml:
```diff
@@ -0,0 +1,55 @@
+  build:
+    strategy:
+      matrix:
+        platform: [ubuntu-latest, windows-latest, macos-latest]
+        node-version: [12, 14]
```
This does node version and cross platform testing out of the box

ChristianMurphy commented on .github/workflows/gh-page.yml:
```diff
@@ -0,0 +1,45 @@
+      - name: Deploy to GitHub pages
+        uses: JamesIves/github-pages-deploy-action@releases/v3
+        with:
+          ACCESS_TOKEN: ${{ secrets.ACCESS_TOKEN }}
```
This deploys storybook to GitHub packages, it does take a bit of setup, it needs an access token to push changes to a branch.

ChristianMurphy commented:
This would also resolve several typing related issues such as: #261 #251 #341 #417

ZupoLlask commented:
When is this fix going to be released?

gilisho commented:
@rexxars sorry for tagging, but I would be really happy if this could be merged, so I'm just pinging.

fivenp commented:
@ChristianMurphy – Sorry for tagging but I'm curious about the progress in here – Is there anything we could help to bring this alive? :)

ChristianMurphy commented:
@fivenp if/when this gets released is up to `@rexxars`.
If you want to use it in the meantime, there is a mirror of this on both github https://github.com/ChristianMurphy/react-remark and npm https://www.npmjs.com/package/@christianmurphy/react-remark

wooorm commented:
@ChristianMurphy Thoughts on the failures on Travis? https://travis-ci.org/github/remarkjs/react-markdown/builds/688158047

wooorm commented:
Hi, thanks for your patience!
I’m gonna go with https://github.com/remarkjs/react-markdown/pull/428, which will result in lots of merge conflicts that it’s probably better to close this and take a new look if there still is something missing.

ChristianMurphy commented:
> I’m gonna go with #428 

:raised_eyebrow: this is #428 

> which will result in lots of merge conflicts that it’s probably better to close this and take a new look if there still is something missing.

I'll take a look at the merge conflicts soon, I can open a fresh PR if preferred.

wooorm commented:
whooops haha

wooorm commented on package.json:
```diff
@@ -1,88 +1,80 @@
-    "unist-util-visit": "^1.4.1",
-    "xtend": "^4.0.1"
+    "rehype-react": "^5.0.1",
+    "remark-parse": "^8.0.2",
```
The plan earlier is explained here: https://github.com/remarkjs/react-markdown/issues/470#issue-717406622.
It seems this PR would be extremely incompatible with the current project. So the plan was instead to mostly bump remark and fix plugins (https://github.com/remarkjs/react-markdown/pull/425) first. Then do this PR later.

hypirion replied:
Ah, thanks, I didn't see that one. Thanks for the quick reply and the good work here!

stjepangolemac commented:
What happened with this?

ChristianMurphy commented:
It was closed by a quirk of GiHub, it was pointed to the v5 branch, which was merged, orphaning this PR, which GH autocloses.

An updated version of this addressing the comments from https://github.com/remarkjs/react-markdown/issues/427#issuecomment-634417637 and https://github.com/remarkjs/react-markdown/issues/427#issuecomment-634865410 is in the works, in the mean time the lower level API that it would build on is already available in https://github.com/ChristianMurphy/react-remark, the next proposal would be a wrapper/preset for react-remark, which matches the existing API more closely.

devuxer commented:
I came to this pull request from #397, and I'm finding myself a little confused about the status of it. I just want my headings to have IDs so I can link to them. How can I do this today with the current version of react-markdown?

wooorm commented:
Still working on it: https://github.com/remarkjs/react-markdown/issues/549

devuxer commented:
@wooorm , Thanks for the quick reply. and I appreciate your work on this. Does #549 mean it is simply not possible to do what I'm trying to do right now with react-markdown, or is there a stopgap/workaround I can throw into my code for now until the next version is released?

Not sure if this is related to #549 as well, but I'm noticing that markdown tables are being converted to HTML tables. I haven't looked yet to see if there's a plug-in for tables, though, so perhaps that would solve it.

wooorm commented:
You can overwrite the heading renderer right now. But 549 indeed means that and is about fixing all this stuff.

Tables are in remark-gfm and work

devuxer commented:
@wooorm , Got it, thanks! Since sending my last message, I decided to give markdown-to-jsx a try instead, and that seems to do what I need out of the box. So, I'll go with that until #549 comes along, then I can reevaluate which library works better for my use case.
